src/OpenSpaceToolkit/Astrodynamics/Trajectory/Segment.cpp

```diff
@@ -97,22 +97,65 @@ Real Segment::Solution::computeDeltaV(const Real& aSpecificImpulse) const
 
 Mass Segment::Solution::computeDeltaMass() const
 {
-    if (segmentType != Segment::Type::Maneuver)
+    if (this->segmentType != Segment::Type::Maneuver)
     {
         return Mass::Kilograms(0.0);
     }
 
     return Mass::Kilograms(getInitialMass().inKilograms() - getFinalMass().inKilograms());
 }
 
+Array<State> Segment::Solution::reComputeStatesAt(
+    const Array<Instant>& anInstantArray, const NumericalSolver& aNumericalSolver
+) const
+{
+    if (this->states.isEmpty())
+    {
+        throw ostk::core::error::RuntimeError("No states exist within segment.");
+    }
+
+    if (anInstantArray.isEmpty())
+    {
+        return Array<State>::Empty();
+    }
+
+    for (Size k = 0; k < anInstantArray.getSize() - 1; ++k)
+    {
+        if (anInstantArray[k] > anInstantArray[k + 1])
+        {
+            throw ostk::core::error::runtime::Wrong("Unsorted Instant Array");
+        }
+    }
+
+    if ((accessStartInstant() > anInstantArray.accessLast()) || (accessEndInstant() < anInstantArray.accessFirst()))
+    {
+        throw ostk::core::error::RuntimeError(
+            "Trying to calculate state outside of segment bounds. No solution available."
+        );
+    }
+
+    const Propagated propagated = {
+        {
+            aNumericalSolver,
+            this->dynamics,
+        },
+        this->states,
+    };
+
+    // TBI: Implement proper caching during dynamics contribution observing MR
+    // this->states = propagated.calculateStatesAt(anInstantArray);
```

Comment:
I don't believe we want to overwrite the observed states... we will start mixing states generated by different numericalSolvers which might not be ideal.